symengine/symengine_exception.h

```diff
@@ -0,0 +1,34 @@
+#ifndef SYMENGINE_EXCEPTION_H
+#define SYMENGINE_EXCEPTION_H
+
+enum symengine_exceptions_t {
+    NO_EXCEPTION,
+    DIV_BY_ZERO,
+    NOT_IMPLEMENTED,
+    UNDEFINED
+};
+
+namespace SymEngine
+{
+
+class SymEngineException : public std::exception
+{
+    std::string m_msg;
+    symengine_exceptions_t ec;
+
+public:
+    SymEngineException(const std::string &msg, symengine_exceptions_t error)
```

Comment:
Instead of passing it as an argument like this, why not to subclass `SymEngineException` with exception classes like DivisionByZero, NotImplemented, Undefined. And make `error_code` virtual, that way the CWrapper macros can stay unchanged (you just check for `SymEngineException` and call `error_code`, just like before). But in the C++ code, you can catch a particular exception. For example in the test, you can test the actual exception, not just the general `SymEngineException`, which can be anything.

I think just like in Python, one should try to catch the most particular exception.